Feature: Optimize ZFS instance creation with image variants (#17630)
canonical/lxd: doc/reference/storage_zfs.md

    zpool set autotrim=on ZPOOL-NAME
    zpool trim ZPOOL-NAME

(storage-zfs-image-variants)=
### Image variants and optimized instance creation

> The ZFS driver supports optimized image volumes that significantly reduces instance creation time.

Reviewer: ```suggestion
The ZFS driver supports optimized image volumes that significantly reduce instance creation time.
```

Author: Hi @Minae, thanks for the thorough review :heart_decoration:. I've updated my commits and reworked the documentation based on your suggestions. Let me know if there's anything else you'd like me to adjust.

When creating an instance from an image, if the image has been previously unpacked into an optimized image volume then LXD can clone the instance from the optimized image volume rather than unpacking from scratch, making instance creation nearly instantaneous.

> The ZFS driver supports having multiple variants of the same optimized image volume in order to support optimized instance creation when using instance specific initial root disk settings (see {ref}`devices-disk-initial-config`).

Reviewer: ```suggestion
If an image is unpacked into an optimized image volume, when LXD creates an instance from that image, it can clone the instance from that volume rather than unpacking from scratch. This makes instance creation nearly instantaneous.

The ZFS driver can maintain multiple variants of the same optimized image volume to support fast instance creation when using instance-specific initial root disk settings (see {ref}`devices-disk-initial-config`).
```

> #### How image variants work

Reviewer: ```suggestion
#### Variant types
```

#### Variant types

LXD can maintain multiple optimized image volumes for the same image, representing different volume configurations:

- **Dataset variant** (dataset mode): A `ZFS filesystem` dataset created when {config:option}`storage-zfs-volume-conf:zfs.block_mode` is `false`.
- **Block-backed variants**: ZFS volumes with specific filesystems (ext4, btrfs, xfs) created when {config:option}`storage-zfs-volume-conf:zfs.block_mode` is `true`.

#### Variant lifecycle

Variants are created lazily on demand when an instance requests a configuration that doesn't match an existing variant.

A variant is deleted when one of the following events occurs:

- The variant is not used by any instance and the last instance that uses it is removed.
- The variant is not used by any instance and the image is deleted.

> - The variant is not used by any instance and the pool configuration is changed such that the variant no longer matches the new pool defaults (for example, a change from dataset mode to block mode).

Reviewer 2: I think maybe I wasn't clear in my last comment. I would retain "The variant is not used by any instance" for the second and third items, but it isn't necessary in the first sentence---once you remove the last instance using the variant, the variant will not be in use anymore:

```suggestion
- The last instance that uses the variant is removed.
- The variant is not used by any instance and the image is deleted.
- The variant is not used by any instance and the pool configuration is changed such that the variant no longer matches the new pool defaults (for example, a change from dataset mode to block mode).
```

Author: Updated the variant cleanup list to remove the redundant "not used by any instance" prefix from the first bullet.

> - Variants with active instances are preserved regardless of configuration changes.

Reviewer: ```suggestion
The following events trigger the deletion of a variant:

- The last instance using the variant is removed.
- No instances are using that variant, and:
  - An image is manually deleted.
  - Pool configuration changes (for example, switching from dataset mode to block mode) and the variant no longer matches the new pool defaults.

Variants with active instances are always preserved.
```

#### Configuration

> To optimize instance creation, ensure your instances use configurations that match your pool defaults:

Reviewer: ```suggestion
To optimize instance creation, use instance configurations that match your pool defaults:
```

- **Pool default**: Set {config:option}`storage-zfs-volume-conf:zfs.block_mode` and {config:option}`storage-zfs-volume-conf:block.filesystem` at the pool level (using the `volume.` prefix) at pool creation.

> - **Instance override**: Use `initial.zfs.block_mode` and `initial.block.filesystem` device configuration for instances requiring different settings.

Reviewer: ```suggestion
- **Instance override**: Use `initial.zfs.block_mode` and `initial.block.filesystem` {ref}`device configuration <devices-disk-initial-config>` for instances requiring different settings.
```

- **Pool default**: Set {config:option}`storage-zfs-volume-conf:zfs.block_mode` and {config:option}`storage-zfs-volume-conf:block.filesystem` at the pool level during pool creation by using the `volume.` prefix.
- **Instance override**: Use `initial.zfs.block_mode` and `initial.block.filesystem` {ref}`device configuration <devices-disk-initial-config>` for instances that require different settings.

#### Performance benefits

Using image variants provides:

> - Faster instance creation (2-3 seconds vs. 10-30 seconds for traditional unpacking).

Reviewer 2: Following up on the comments by @simondeziel , you could remove the number of seconds if the time to create an instance might vary and possibly exceed 3 seconds under some circumstances:

Simplest is probably:

```suggestion
- Faster instance creation.
```

Alternatively, based on @simondeziel 's comments, something along the lines of the following:

```suggestion
- Faster instance creation (the time to clone a volume with ZFS is unrelated to the volume size).
```

Reviewer 3: Both versions LGTM, thanks!

Author: Replaced the timing figures with the accurate ZFS semantics (clone time is size-independent). Thanks :sparkle:

    zpool set autotrim=on ZPOOL-NAME
    zpool trim ZPOOL-NAME

(storage-zfs-image-variants)=
### Image variants and optimized instance creation

The ZFS driver supports optimized image volumes that significantly reduce instance creation time.

If an image is unpacked into an optimized image volume, when LXD creates an instance from that image, it can clone the instance from that volume rather than unpacking from scratch.

> This makes instance creation nearly instantaneous.

Reviewer 2: Avoid hyperbolic claims. How about:

```suggestion
This speeds up instance creation.
```

The ZFS driver can maintain multiple variants of the same optimized image volume to support fast instance creation when using instance-specific initial root disk settings (see {ref}`devices-disk-initial-config`).

#### Variant types

LXD can maintain multiple optimized image volumes for the same image, representing different volume configurations:

- **Dataset variant** (dataset mode): A `ZFS filesystem` dataset created when {config:option}`storage-zfs-volume-conf:zfs.block_mode` is `false`.
- **Block-backed variants**: ZFS volumes with specific filesystems (ext4, btrfs, xfs) created when {config:option}`storage-zfs-volume-conf:zfs.block_mode` is `true`.

#### Variant lifecycle

> Variants are created lazily on-demand when an instance requests a configuration that doesn't match an existing variant.

Reviewer 2: Nitpicking here, but "on-demand" only has a hyphen when it is used as an adjective (e.g. "on-demand creation"). Here, this should be "on demand".

```suggestion
Variants are created lazily on demand when an instance requests a configuration that doesn't match an existing variant.
```

A variant is deleted when one of the following events occurs:

- The last instance that uses the variant is removed.
- The image is deleted and no instances use the variant.

> - The pool configuration changes (for example, switching from dataset mode to block mode) and the variant no longer matches the new pool defaults and has no active instances.

Reviewer 2: How about moving the "no instances use the variant" part to the beginning of both sentences to emphasize this important, shared condition? I've also moved your example in the final sentence to the end of that sentence to keep the information about the event together:

```suggestion
- The variant is not used by any instance and the image is deleted.
- The variant is not used by any instance and the pool configuration is changed such that the variant no longer matches the new pool defaults (for example, a change from dataset mode to block mode).
```

Variants with active instances are always preserved regardless of image deletion or configuration changes.

#### Configuration

To optimize instance creation, use instance configurations that match your pool defaults:

- **Pool default**: Set {config:option}`storage-zfs-volume-conf:zfs.block_mode` and {config:option}`storage-zfs-volume-conf:block.filesystem` at the pool level (using the `volume.` prefix) at pool creation.
- **Instance override**: Use `initial.zfs.block_mode` and `initial.block.filesystem` {ref}`device configuration <devices-disk-initial-config>` for instances requiring different settings.

> Changing pool configuration frequently can reduce optimization benefits, as variants may need to be recreated.

Reviewer 2: What you've written in the first sentence of this section is that users should match instance configurations to the pool defaults, but what you describe in the list is the way for users to set the pool defaults and then set up *different* settings for instances.

What I think you want to communicate is: (1) changing the pool configuration limits users' ability to optimize instance creation; (2) to optimize, use pool configurations that will be shared with most instances (i.e. plan ahead); (3) it will still be possible to set up alternative instance configurations when needed.

If I have that right, I'm wondering if it might help to move the problem (frequent changes to pool configuration) to the front, and then provide the solution (pool configuration, with the option to override). When providing the solution, we can also clarify that the goal is to set up pool defaults that will be used by most instances. How about the following (with a little additional rewording):

```suggestion
Changing pool configuration frequently can reduce optimization benefits, as variants may need to be recreated.
To optimize instance creation, set up default pool configurations and create instances that use those defaults (you can still override those settings for specific instances as needed):

- **Pool default**: Set {config:option}`storage-zfs-volume-conf:zfs.block_mode` and {config:option}`storage-zfs-volume-conf:block.filesystem` at the pool level during pool creation by using the `volume.` prefix.
- **Instance override**: Use `initial.zfs.block_mode` and `initial.block.filesystem` {ref}`device configuration <devices-disk-initial-config>` for instances that require different settings.
```

Author: Hi @elijahgreenstein, thank you so much for taking the time to review the documentation and provide such thoughtful feedback! Your suggestions on tone, grammar, and clarity significantly improved the quality of the docs. I particularly appreciated your guidance on removing hyperbolic language and restructuring the Configuration section to better emphasize the planning aspect, it really does make the content more accessible and practical for users.

I've incorporated all of your suggestions into the commits. Thanks again for the meticulous review!

Reviewer 2: Hi @tugbataluy , great, I'm glad you found the suggestions helpful!

#### Performance benefits

Using image variants provides:

> - Near-instantaneous instance creation (2-3 seconds vs. 10-30 seconds for traditional unpacking).

Reviewer 2: Avoid hyperbolic language.

```suggestion
- Faster instance creation (2-3 seconds vs. 10-30 seconds for traditional unpacking).
```

Reviewer 3: For me, near-instantaneous was less controversial than the seconds figures :)

ZFS being a CoW filesystem, creating a clone is really fast and the time it takes is not related to the size of the original volume (the clone source).

On the other hand, when doing a traditional (no CoW) unpacking, the time it takes is directly proportional to the size of the original volume as the data needs to be moved around on disk.

Providing time estimates is thus really hard: what if I have a multi-GiB image? what if I'm unfortunate to use a HDD? However with ZFS and it's CoW, in both case the time needed to create a clone is pretty similar and close to instantaneous ;)

But yeah, in general think it's good to avoid hyperbolic claims :)

    zpool set autotrim=on ZPOOL-NAME
    zpool trim ZPOOL-NAME

(storage-zfs-image-variants)=
### Image variants and optimized instance creation

The ZFS driver in LXD implements an optimization mechanism called *image variants* that significantly improves instance creation performance.
When creating an instance from an image, LXD can clone the instance from a pre-unpacked image variant instead of unpacking the image from scratch, making instance creation nearly instantaneous.

#### How image variants work

LXD maintains multiple versions of the same image with different storage configurations:

- **Base variant** (dataset mode): A `ZFS filesystem` dataset created when {config:option}`storage-zfs-volume-conf:zfs.block_mode` is `false`.
- **Block-backed variants**: ZFS volumes with specific filesystems (ext4, btrfs, xfs) created when {config:option}`storage-zfs-volume-conf:zfs.block_mode` is `true`.

Each variant is stored with a naming convention: `<pool>/images/<fingerprint>` for the base variant, and `<pool>/images/<fingerprint>_<filesystem>` for block-backed variants (for example, `pool/images/abc123_ext4`).

> #### Variant lifecycle

Reviewer 4: This is describing the implementation, so I think would be best moved to the Internals doc section, as is not relevant for the normal user.
